provisioning/cmd/provisioning-service/main.go

```diff
@@ -4,83 +4,93 @@
-		ListenAddr string `mapstructure:"listen-addr"`
-	}
-	VtsGRPC *viper.Viper
+type cfg struct {
+	PluginDir  string `mapstructure:"plugin-dir"`
+	ListenAddr string `mapstructure:"listen-addr" valid:"dialstring"`
 }
 
-func initConfig() (*config, error) {
-	v := viper.New()
+func (o cfg) Validate() error {
+	if _, err := os.Stat(o.PluginDir); err != nil {
+		return fmt.Errorf("could not stat PluginDir: %w", err)
+	}
 
-	v.SetConfigType("yaml")
-	v.SetConfigName("config")
+	return nil
+}
 
-	wd, err := os.Getwd()
+func main() {
+	v, err := config.ReadRawConfig("", false)
 	if err != nil {
-		return nil, err
+		log.Fatalf("Could not read config sources: %v", err)
 	}
-	v.AddConfigPath(wd)
 
-	if err := v.ReadInConfig(); err != nil {
-		return nil, err
+	cfg := cfg{
+		PluginDir:  DefaultPluginDir,
+		ListenAddr: DefaultListenAddr,
 	}
 
-	v.SetDefault("provisioning.plugin-dir", DefaultPluginDir)
-	v.SetDefault("provisioning.listen-addr", DefaultListenAddr)
-
-	var cfg config
+	subs, err := config.GetSubs(v, "provisioning", "*vts", "*logging")
+	if err != nil {
+		log.Fatal(err)
+	}
 
-	if err = v.UnmarshalKey("provisioning", &cfg.Provisioning); err != nil {
-		return nil, err
+	classifiers := map[string]interface{}{"service": "provisioning"}
+	if err := log.Init(subs["logging"], classifiers); err != nil {
+		log.Fatalf("could not configure logging: %v", err)
 	}
+	log.InitGinWriter() // route gin output to our logger.
 
-	if cfg.VtsGRPC = v.Sub("vts-grpc"); cfg.VtsGRPC == nil {
-		return nil, errors.New(`"vts-grpc" section not found in config.`)
+	loader := config.NewLoader(&cfg)
+	if err = loader.LoadFromViper(subs["provisioning"]); err != nil {
+		log.Fatalf("Could not load config: %v", err)
 	}
 
-	return &cfg, nil
-}
+	log.Info("loading plugins")
+	pluginManager := NewGoPluginManager(cfg.PluginDir, log.Named("plugin"))
 
-func main() {
-	cfg, err := initConfig()
-	if err != nil {
-		log.Fatalf("could not load config: %v", err)
+	log.Info("initializing VTS client")
+	vtsClient := vtsclient.NewGRPC()
+	if err := vtsClient.Init(subs["vts"]); err != nil {
+		log.Fatalf("Could not initilize VTS client: %v", err)
 	}
 
-	pluginDir := cfg.Provisioning.PluginDir
-	listenAddr := cfg.Provisioning.ListenAddr
+	vtsServerVersion, err := vtsClient.GetVersion(context.TODO(), &emptypb.Empty{})
+	if err == nil {
+		log.Infow("vts connection established", "server-version", vtsServerVersion.Version)
+	} else {
+		log.Warnw("Could not connect to VTS server. If you do not expect the server to be running yet, this is probably OK, otherwise it may indicate an issue with your vts.server-addr in your settings",
+			"error", err)
+	}
 
-	pluginManager := NewGoPluginManager(pluginDir)
-	vtsClient := vtsclient.NewGRPC(cfg.VtsGRPC)
-	apiHandler := api.NewHandler(pluginManager, vtsClient)
-	go apiServer(apiHandler, listenAddr)
+	log.Infow("initializing API service", "address", cfg.ListenAddr)
+	apiHandler := api.NewHandler(pluginManager, vtsClient, log.Named("api"))
```

Comment:
we have two different API services, maybe we should qualify which "api" is speaking here